### PR TITLE
feat: add `ml/strided/dsgd-trainer`

### PR DESCRIPTION
**Pull request by nakul-krishnakumar**
Resolves a part of  #12875.

## Description

> What is the purpose of this pull request?

This pull request:

-   Adds `ml/strided/dsgd-trainer`.

## Related Issues

> Does this pull request have any related issues?

This pull request has the following related issues:

-   #12875.

## Questions

> Any questions for reviewers of this pull request?

No.

## Other

> Any other information relevant to this pull request? This may include screenshots, references, and/or implementation notes.

No.

## Checklist

> Please ensure the following tasks are completed before submitting this pull request.

-   [x] Read, understood, and followed the [contributing guidelines][contributing].

### AI Assistance

> When authoring the changes proposed in this PR, did you use any kind of AI assistance?

-   [x] Yes
-   [ ] No

If you answered "yes" above, how did you use AI assistance?

-   [x] Code generation (e.g., when writing an implementation or fixing a bug)
-   [ ] Test/benchmark generation
-   [ ] Documentation (including examples)
-   [x] Research and understanding

#### Disclosure

> If you answered "yes" to using AI assistance, please provide a short disclosure indicating how you used AI assistance. This helps reviewers determine how much scrutiny to apply when reviewing your contribution. Example disclosures: "This PR was written primarily by Claude Code." or "I consulted ChatGPT to understand the codebase, but the proposed changes were fully authored manually by myself.".

I used both Claude Code and ChatGPT to help me refer the sklearn implementation and write the code.

* * *

@stdlib-js/reviewers

[contributing]: https://github.com/stdlib-js/stdlib/blob/develop/CONTRIBUTING.md

### CODE REVIEW

**Comment by nakul-krishnakumar** [comment]
I have the following doubts:
1. As we are anyway having branching for `penalties` and `learningRate`, I think it would be okay to keep a single API `ml/strided/sgd-trainer` with branching for `lossFunctions` as well as instead of having separate sgd trainers per loss function as we had planned on in the earlier meetings (`dsgd-trainer-squared-epsilon-insensitive`).
2. Should we implement the helper function `l1Penalty` as a standalone implementation? ( I don't see any reason why we shouldn't )
3. Should the API support all 3 penalties/regularization? (`l1`, `l2` and `elasticnet`) 
     - `ml/incr/binary-classification` supports only `L2` by default.
     - If going for all 3, we will have to implement standalone `ml/base/sgd-classification/penalties` (enum) (opened a PR #13623 )
     - I would suggest to have support for all 3.
4. How do we handle returning `intercept`? We cant just update the user passed `intercept` as we do for the `weight matrix` as I believe numbers are not passed by reference unlike arrays.
5. Currently this implementation is not in 1:1 parity with the sklearn API, it skips things like dataset shuffling, learning rate schedulers like `PA1`, `PA2` (they were recently added in sklearn API, so I am skeptical on adding those), early stopping etc. It would be helpful if you could let me know if I am missing anything important here that you think would be helpful to have.
6. Should we have a weight vector implementation similar to how [sklearn](https://github.com/scikit-learn/scikit-learn/blob/89247fcd7f02bbd92c9da141250ec8b82caf8dd0/sklearn/utils/weight_vector.pyx) as well as [`ml/incr/sgd-regression`](https://github.com/stdlib-js/stdlib/blob/develop/lib/node_modules/%40stdlib/ml/incr/sgd-regression/lib/weight_vector.js) does?
sklearn using weight vector in sgd implementation: [Ref](https://github.com/scikit-learn/scikit-learn/blob/fe2edb3cdbd75ae4e662fda67dcb19277258792b/sklearn/linear_model/_sgd_fast.pyx.tp#L411)

7. Tidy up function argument, what all should be passed to base implementation?

sklearn sgd implementation: [Ref](https://github.com/scikit-learn/scikit-learn/blob/fe2edb3cdbd75ae4e662fda67dcb19277258792b/sklearn/linear_model/_sgd_fast.pyx.tp)

cc: @kgryte

**Comment by nakul-krishnakumar** [comment]
Also please note that I kept some comments only for the sake of understanding, will remove them once we finalize the implementation.

**Comment by kgryte** [comment]
Quick aside: I am wondering if we should namespace the strided routines. E.g.,

```
ml/strided/kmeans/ddistance
ml/strided/sgd-trainer/dsquared-epsilon-insensitive
...
```

The current package base name, as is, is a bit unwieldy. Thoughts?

**Comment by kgryte** [comment]
For packages which are applicable across ML algos, they can go directly in `ml/strided`. E.g., hypothetically,

```
ml/strided/dsquared-epsilon-insensitive
```

And actually, is squared epsilon insensitive so specific to sgd-trainer that it shouldn't be broadly applicable?

I am just curious where we draw the line.

**Comment by nakul-krishnakumar** [comment]
> Quick aside: I am wondering if we should namespace the strided routines. E.g.,
> 
> ```
> ml/strided/kmeans/ddistance
> ml/strided/sgd-trainer/dsquared-epsilon-insensitive
> ...
> ```
> 
> The current package base name, as is, is a bit unwieldy. Thoughts?

Yeah we could do that, we had discussed this earlier but decided not to until we add more packages and it becomes harder to track.

**Comment by nakul-krishnakumar** [comment]
> For packages which are applicable across ML algos, they can go directly in `ml/strided`. E.g., hypothetically,
> 
> ```
> ml/strided/dsquared-epsilon-insensitive
> ```
> 
> And actually, is squared epsilon insensitive so specific to sgd-trainer that it shouldn't be broadly applicable?
> 
> I am just curious where we draw the line.

Here, yes it is specific to `SGD`. Because `ml/strided/dsquared-epsilon-insensitive` is not actually a strided version of  `squared-epsilon-insensitive` loss function gradient, instead it is a strided version of `SGD` implementation hard coded with `squared-epsilon-insensitive` loss function.
Instead what I am saying is lets just have a single `ml/strided/dsgd-trainer` which uses a package similar to `ml/strided/dkmeans-distance` and resolve the gradient of the user passed loss function.
```javascript
function dsgdTrainer( loss, .... ) {
    ...
    g = dsgdLossGradient( loss, x, e, y, p );
    ...
}
```
This was our first choice but decided on implementing individual `sgd-trainer` per loss function so that we can optimize it (as this doesn't require resolving the loss function using branching). But now that I look at it, the implementation anyway does branching to resolve `learning rate` and `penalty` so maybe branching for resolving `loss functions` too might not be that costly comparatively.

cc: @kgryte

File: lib/node_modules/@stdlib/ml/strided/dsgd-trainer/lib/learning_rate.js [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+
+// FUNCTIONS //
+
+/**
+* Computes learning rate by applying the basic learning rate scheduler.
+*
+* Note:
+*
+* -   Here `params` => `[ ]` (empty)
```
Comment (nakul-krishnakumar) [inline lib/node_modules/@stdlib/ml/strided/dsgd-trainer/lib/learning_rate.js]:
```suggestion
* -   Here `params` => `[ eta0 ]` (empty)
```

File: lib/node_modules/@stdlib/ml/strided/dsgd-trainer/lib/learning_rate.js [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+* -   Here `params` => `[ lambda ]`
+*
+* @private
+* @param {NonNegativeInteger} t - current iteration.
+* @param {Float64Array} params - strided array containing scheduler specific parameters.
+* @returns {number} learning rate
+*/
+function pegasos( t, params ) {
+	return 1.0 / ( params[ 0 ]*t );
+}
```
Comment (nakul-krishnakumar) [inline lib/node_modules/@stdlib/ml/strided/dsgd-trainer/lib/learning_rate.js]:
```suggestion
* -   Here `params` => `[ eta0, lambda ]`
*
* @private
* @param {NonNegativeInteger} t - current iteration.
* @param {Float64Array} params - strided array containing scheduler specific parameters.
* @returns {number} learning rate
*/
function pegasos( t, params ) {
	return 1.0 / ( params[ 1 ]*t );
}
```

File: lib/node_modules/@stdlib/ml/strided/dsgd-trainer/lib/decay.js [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+* -   Here `params` => `[ eta, lambda, l1Ratio ]`
+*
+* @private
+* @param {NonNegativeInteger} scaleFactor - current iteration.
+* @param {Float64Array} params - strided array containing regularizer specific parameters.
+* @returns {number} scale factor
+*/
+function l2Decay( scaleFactor, params ) {
+	return scaleFactor * max( 0.0, 1.0 - ( ( 1.0 - params[ 2 ] ) * params[ 0 ] * params[ 1 ] ) ); // eslint-disable-line max-len
+}
```
Comment (nakul-krishnakumar) [inline lib/node_modules/@stdlib/ml/strided/dsgd-trainer/lib/decay.js]:
```suggestion
* -   Here `params` => `[ lambda, l1Ratio ]`
*
* @private
* @param {NonNegativeInteger} scaleFactor - current iteration.
* @param {Float64Array} params - strided array containing regularizer specific parameters.
* @returns {number} scale factor
*/
function l2Decay( scaleFactor, eta, params ) {
	return scaleFactor * max( 0.0, 1.0 - ( ( 1.0 - params[ 2 ] ) * eta * params[ 1 ] ) ); // eslint-disable-line max-len
}
```

File: lib/node_modules/@stdlib/ml/strided/dsgd-trainer/lib/decay.js [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+* -   Here `params` => `[ ]` (empty)
+*
+* @private
+* @param {NonNegativeInteger} scaleFactor - current iteration.
+* @param {Float64Array} params - strided array containing regularizer specific parameters.
+* @returns {number} scale factor
+*/
+function identityDecay( scaleFactor, params ) { // eslint-disable-line no-unused-vars
+	return scaleFactor;
+}
```
Comment (nakul-krishnakumar) [inline lib/node_modules/@stdlib/ml/strided/dsgd-trainer/lib/decay.js]:
```suggestion
function identityDecay( scaleFactor, eta, params ) { // eslint-disable-line no-unused-vars
	return scaleFactor;
}
```

File: lib/node_modules/@stdlib/ml/strided/dsgd-trainer/lib/base.js [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+	*    - Should we implement a standalone weight matrix just like how `ml/incr/sgd-regression` as well as
+	*        the sklearn API did? They referred sofia-ml as far as I know.
+	*/
+
+	// Do we set this here or expect higher level user to pass it?
+	if ( penalty === 'l2' ) {
+		l1Ratio = 0.0;
+	} else if ( penalty === 'l1' ) {
+		l1Ratio = 1.0;
+	}
```
Comment (nakul-krishnakumar) [inline lib/node_modules/@stdlib/ml/strided/dsgd-trainer/lib/base.js]:
```suggestion
	if ( penalty === 'l2' ) {
		penaltyParams[ 1 ] = 0.0;
	} else if ( penalty === 'l1' ) {
		penaltyParams[ 1 ] = 1.0;
	}
```

File: lib/node_modules/@stdlib/ml/strided/dsgd-trainer/include/stdlib/ml/strided/dsgd_trainer.h [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+* If C++, prevent name mangling so that the compiler emits a binary file having undecorated names, thus mirroring the behavior of a C compiler.
+*/
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+* Struct for storing SGD parameters.
+*/
+struct stdlib_ml_sgd_params_float64_params {
```
Comment (kgryte) [inline lib/node_modules/@stdlib/ml/strided/dsgd-trainer/include/stdlib/ml/strided/dsgd_trainer.h]:
```suggestion
struct stdlib_ml_sgd_float64_params {
```

File: lib/node_modules/@stdlib/ml/strided/dsgd-trainer/test/fixtures/python/runner.py [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+    gen("./loss-functions/loss_squared_epsilon_insensitive.json",
+        loss="squared_epsilon_insensitive")
+    gen("./loss-functions/loss_huber.json", loss="huber")
+
+    # Loss functions (binary classification, `y` in `{-1,+1}`):
+    gen("./loss-functions/loss_hinge.json", loss="hinge")
+    gen("./loss-functions/loss_squared_hinge.json", loss="squared_hinge")
+    gen("./loss-functions/loss_perceptron.json", loss="perceptron")
+    gen("./loss-functions/loss_modified_huber.json", loss="modified_huber")
+    gen("./loss-functions/loss_log.json", loss="log_loss")
```
Comment (kgryte) [inline lib/node_modules/@stdlib/ml/strided/dsgd-trainer/test/fixtures/python/runner.py:230]: Is there a reason to use `./` here and not in the non-sub-directory file paths?
